NYPLAudiobookToolkit/Utils/MediaProcessor.swift

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+        fh.closeFile()
+        return atoms
+    }
+    
+    private static func getAtoms(data: Data, offset: UInt64) -> [QTAtomMetadata] {
+        var localOffset: UInt64 = offset + 8
+        var atoms: [QTAtomMetadata] = []
+        
+        while localOffset < data.count {
+            var size: UInt64
+            do {
+                size = try UInt64(data.bigEndianUInt32At(offset: Int(localOffset)))
+            } catch {
+                print("Could not read atom size")
+                atoms = []
+                break
+            }
+            
+            let type = String(data: data.subdata(in: Range(Int(localOffset+4)...Int(localOffset+7))), encoding: .ascii) ?? ""
+            if size == 1 {
+                do {
+                    size = try data.bigEndianUInt64At(offset: Int(localOffset+8))
+                } catch {
+                    print("Could not read atom ext size")
+                    atoms = []
+                    break
+                }
+            }
+            atoms.append(QTAtomMetadata(offset: localOffset, size: size, type: type))
+            localOffset += size
+        }
+        return atoms
+    }
+    
+    private static func patchMoovData(data: inout Data, moov: QTAtomMetadata) -> Bool {
+        let moovChildren = getAtoms(data: data, offset: 0)
+        guard let trakAtom = moovChildren.first(where: { $0.type == "trak" }) else {
+            ATLog(.warn, "No trak atom found")
+            return false
+        }
+        
+        let trakChildren = getAtoms(data: data, offset: trakAtom.offset)
+        guard let mdiaAtom = trakChildren.first(where: { $0.type == "mdia" }) else {
+            ATLog(.warn, "No mdia atom found")
+            return false
+        }
+        
+        let mdiaChildren = getAtoms(data: data, offset: mdiaAtom.offset)
+        guard let minfAtom = mdiaChildren.first(where: { $0.type == "minf" }) else {
+            ATLog(.warn, "No minf atom found")
+            return false
+        }
+        
+        let minfChildren = getAtoms(data: data, offset: minfAtom.offset)
+        guard let stblAtom = minfChildren.first(where: { $0.type == "stbl" }) else {
+            ATLog(.warn, "No stbl atom found")
+            return false
+        }
+        
+        let stblChildren = getAtoms(data: data, offset: stblAtom.offset)
+        for c in stblChildren {
+            if c.type == "stco" || c.type == "co64" {
+                do {
+                    try patchChunkOffsetAtom(data: &data, atom: c, moovSize: Int(moov.size))
+                } catch {
+                    ATLog(.warn, "Error patching chunk offset atom. \(error)")
+                    return false
+                }
+            }
+        }
+        return true
+    }
+    
+    private static func patchChunkOffsetAtom(data: inout Data, atom: QTAtomMetadata, moovSize: Int) throws {
+        let entryCount = try data.bigEndianUInt32At(offset: Int(atom.offset+12))
```

Comment:
can you explain these magic numbers here? e.g. 12, 16, * 8, * 4 etc below -- not sure what I'm looking at!